Framework/Parallel/inc/MantidParallel/IO/MultiProcessEventLoader.h

```diff
@@ -0,0 +1,306 @@
+                  const std::vector<std::string> &bankNames,
+                  const std::vector<int32_t> &bankOffsets, unsigned from,
+                  unsigned to) {
+  std::vector<int32_t> eventId;
+  std::vector<T> eventTimeOffset;
+
+  std::size_t eventCounter{0};
+  auto bankSizes = EventLoader::readBankSizes(instrument, bankNames);
+  IO::NXEventDataLoader<T> loader(1, instrument, bankNames);
+  for (unsigned bankIdx = 0; bankIdx < bankNames.size(); ++bankIdx) {
+    auto count = bankSizes[bankIdx];
+
+    bool isFirstBank = (eventCounter < from);
+
+    if (eventCounter + count > from) {
+      size_t start{0};
+      size_t finish{count};
+      if (isFirstBank)
+        start = from - eventCounter;
+      if (eventCounter + count > to)
+        finish = to - eventCounter;
+
+      std::size_t cnt = finish - start;
+
+      auto part = loader.setBankIndex(bankIdx);
+      eventTimeOffset.resize(cnt);
+      loader.readEventTimeOffset(eventTimeOffset.data(), start, cnt);
+      eventId.resize(cnt);
+      loader.readEventID(eventId.data(), start, cnt);
+
+      detail::eventIdToGlobalSpectrumIndex(eventId.data(), cnt,
+                                           bankOffsets[bankIdx]);
+
+      std::unordered_map<int32_t, std::size_t> eventsPerPixel;
+      for (auto &pixId : eventId) {
+        auto iter = eventsPerPixel.find(pixId);
+        if (iter == eventsPerPixel.end())
+          iter = eventsPerPixel.insert(std::make_pair(pixId, 0)).first;
+        ++iter->second;
+      }
+
+      for (const auto &pair : eventsPerPixel)
+        storage.reserve(0, pair.first, pair.second);
+
+      part->setEventOffset(start);
+      for (unsigned i = 0; i < eventId.size(); ++i) {
+        try {
+          TofEvent event{(double)eventTimeOffset[i], part->next()};
+          storage.AppendEvent(0, eventId[i], event);
+        } catch (std::exception const &ex) {
+          std::rethrow_if_nested(ex);
+        }
+      }
+    }
+    eventCounter += count;
+    if (eventCounter >= to)
+      break;
+  }
+}
+
+/**Loades the portion of data from hdf5 given group and banks list to shared
+ * memory, 'from' and 'to' represent the range in global number of events.
+ * Implements the 'producer-consumer' strategy*/
+template <>
+template <typename T>
+void MultiProcessEventLoader::GroupLoader<
+    MultiProcessEventLoader::LoadType::producerConsumer>::
+    loadFromGroup(EventsListsShmemStorage &storage, const H5::Group &instrument,
+                  const std::vector<std::string> &bankNames,
+                  const std::vector<int32_t> &bankOffsets, unsigned from,
+                  unsigned to) {
+
+  const std::size_t chLen{std::max<std::size_t>((to - from) / 100, 1)};
+  auto bankSizes = EventLoader::readBankSizes(instrument, bankNames);
+
+  struct Task {
+    std::vector<int32_t> eventId;
+    std::vector<T> eventTimeOffset;
+    IO::NXEventDataLoader<T> loader;
+    decltype(loader.setBankIndex(0)) partitioner;
```

Comment:
It is not, this is only for type deduction. I think this is a good way to define the type of the partitioner, because it is complicated template type.